src/plugins/visualizations/public/plugin.ts

```diff
@@ -18,42 +18,155 @@
+ *
+ * @public
+ */
+
+export type VisualizationsSetup = TypesSetup;
+
+export interface VisualizationsStart extends TypesStart {
+  savedVisualizationsLoader: SavedVisualizationsLoader;
+  createVis: (visType: string, visState?: SerializedVis) => Vis;
+  convertToSerializedVis: typeof convertToSerializedVis;
+  convertFromSerializedVis: typeof convertFromSerializedVis;
+  showNewVisModal: typeof showNewVisModal;
+}
 
 export interface VisualizationsSetupDeps {
   expressions: ExpressionsSetup;
+  embeddable: EmbeddableSetup;
+  usageCollection: UsageCollectionSetup;
+  data: DataPublicPluginSetup;
 }
 
 export interface VisualizationsStartDeps {
+  data: DataPublicPluginStart;
   expressions: ExpressionsStart;
+  uiActions: UiActionsStart;
 }
 
-export type VisualizationsSetup = void;
-
-export type VisualizationsStart = void;
-
-export class VisualizationsPublicPlugin
+/**
+ * Visualizations Plugin - public
+ *
+ * This plugin's stateful contracts are returned from the `setup` and `start` methods
+ * below. The interfaces for these contracts are provided above.
+ *
+ * @internal
+ */
+export class VisualizationsPlugin
   implements
     Plugin<
       VisualizationsSetup,
       VisualizationsStart,
       VisualizationsSetupDeps,
       VisualizationsStartDeps
     > {
+  private readonly types: TypesService = new TypesService();
+
   constructor(initializerContext: PluginInitializerContext) {}
 
-  public setup(core: CoreSetup, { expressions }: VisualizationsSetupDeps): VisualizationsSetup {
+  public setup(
+    core: CoreSetup,
+    { expressions, embeddable, usageCollection, data }: VisualizationsSetupDeps
+  ): VisualizationsSetup {
+    setUISettings(core.uiSettings);
+    setUsageCollector(usageCollection);
+
+    expressions.registerFunction(visualizationFunction);
+    expressions.registerRenderer(visualizationRenderer);
     expressions.registerFunction(rangeExpressionFunction);
     expressions.registerFunction(visDimensionExpressionFunction);
 
-    return undefined;
+    const embeddableFactory = new VisualizeEmbeddableFactory();
+    embeddable.registerEmbeddableFactory(VISUALIZE_EMBEDDABLE_TYPE, embeddableFactory);
+
+    return {
+      ...this.types.setup(),
+    };
   }
 
-  public start(core: CoreStart, { expressions }: VisualizationsStartDeps): VisualizationsStart {
-    return undefined;
+  public start(
+    core: CoreStart,
+    { data, expressions, uiActions }: VisualizationsStartDeps
+  ): VisualizationsStart {
+    const types = this.types.start();
+    setI18n(core.i18n);
+    setTypes(types);
+    setCapabilities(core.application.capabilities);
+    setHttp(core.http);
+    setSavedObjects(core.savedObjects);
+    setIndexPatterns(data.indexPatterns);
+    setFilterManager(data.query.filterManager);
+    setExpressions(expressions);
+    setUiActions(uiActions);
+    setTimeFilter(data.query.timefilter.timefilter);
+    setAggs(data.search.aggs);
+    setOverlays(core.overlays);
+    setChrome(core.chrome);
```

Comment:
Probably out of the scope of the PR and heavy changes, but imho that kind of module-static service setters/accessors don't make much sense in NP.

Reply:
@pgayvallet Yes, 100% agreed, and this is a discussion we are having now. This pattern was started for a very specific use case, usage grew, and now we are trying to move away from it as it has created weird bugs and more complex test mocks.

The problem it was trying to solve is: how do we access `start` services from within items added to a registry in the same plugin's `setup` method? (i.e. you cannot use `core.getStartServices` as it is not a dependency)